### PR TITLE
syslinux: Use Nixpkgs' gnu-efi

### PR DESCRIPTION
**Pull request by samueldr**
## Description of changes

#316878 aimed to change how `gnu-efi` was fetched. My understanding (after futzing about) is that sourceforge's git endpoint now doesn't provide direct commit access.

Instead of awkwardly fetching an older `gnu-efi`, prefer using Nixpkgs' `gnu-efi`.

Note that *within Nixpkgs* only the "legacy bios" syslinux is used. We could also drop the EFI build instead.

Closes #316878.

## Things done

```
 $ nix-build -A syslinux -A pkgsi686Linux.syslinux
```

Ran the `.efi` output, from before the change, and after. Same result.

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/os-specific/linux/syslinux/default.nix [hunk 1]
```diff
@@ -111,8 +115,22 @@ stdenv.mkDerivation {
     "MANDIR=$(out)/share/man"
     "PERL=perl"
     "HEXDATE=0x00000000"
+    # Works around confusing (unrelated) error messages when upx is not made availab.e
```
Comment (raboof) [inline pkgs/os-specific/linux/syslinux/default.nix:118]:
```suggestion
    # Works around confusing (unrelated) error messages when upx is not made available
```

**Review by raboof — COMMENTED** [review]
Looks great to me, I can't really judge how risky it is to use a different version of gnu-efi than is referenced by the syslinux submodule...

**Comment by samueldr** [comment]
> Looks great to me, I can't really judge how risky it is to use a different version of gnu-efi than is referenced by the syslinux submodule...

The main risk would be that it stops building with it due to backwards-incompatible changes.

Technically the added patch works around a backwards-incompatible change, but *barely* so. (A feature was split off out into its own header.)

If it causes issues down the line, I believe it should be fine to disable building the EFI outputs entirely. Anyways they are not used within Nixpkgs nor NixOS. To do so would be relatively simple, removing the references to `gnu-efi`, and then removing the EFI targets.

I actually hesitated and almost went with just removing the EFI outputs.

**Review by raboof — APPROVED** [review]

**Comment by K900** [comment]
Somehow, and I genuinely have no idea how, this breaks the ISO boot: https://cache.nixos.org/log/z24ffw9730kzaaljzy18g67jjnn2ix9h-vm-test-run-boot-bios-cdrom.drv

**Comment by raboof** [comment]
oof, thanks for the quick revert :/

**Comment by raboof** [comment]
created #317158

File: pkgs/os-specific/linux/syslinux/default.nix [hunk 2]
```diff
@@ -16,11 +18,10 @@ stdenv.mkDerivation {
 
   # This is syslinux-6.04-pre3^1; syslinux-6.04-pre3 fails to run.
   # Same issue here https://www.syslinux.org/archives/2019-February/026330.html
-  src = fetchgit {
-    url = "https://repo.or.cz/syslinux";
-    rev = "b40487005223a78c3bb4c300ef6c436b3f6ec1f7";
-    sha256 = "sha256-GqvRTr9mA2yRD0G0CF11x1X0jCgqV4Mh+tvE0/0yjqk=";
-    fetchSubmodules = true;
+  src = fetchFromRepoOrCz {
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/syslinux/default.nix:21]: Didn't know that existed. fetchgit defaults to fetching submodules, I assume this doesnt

File: pkgs/os-specific/linux/syslinux/default.nix [hunk 3]
```diff
@@ -65,19 +66,21 @@ stdenv.mkDerivation {
       "0018-prevent-pow-optimization.patch"
       "26f0e7b2"
       "sha256-dVzXBi/oSV9vYgU85mRFHBKuZdup+1x1BipJX74ED7E=")
+    (fetchpatch {
+      # Fixes build with "modern" gnu-efi
+      # https://github.com/samueldr/syslinux/commit/68defee52f4eba82eefaeea17f21c7498448dd6b
+      url = "https://github.com/samueldr/syslinux/commit/68defee52f4eba82eefaeea17f21c7498448dd6b.patch";
+      hash = "sha256-5xIqM8Gq8D86HL1i7fBTHRB/ZR2XtCz5VsT4bI5vneo=";
+    })
```
Comment (SuperSandro2000) [inline pkgs/os-specific/linux/syslinux/default.nix:74]: tbh we would have been probably better off to just vendor that tiny patch to not rely on your repo 😅

File: pkgs/os-specific/linux/syslinux/default.nix (hunk 2 above)
Comment (samueldr) [reply pkgs/os-specific/linux/syslinux/default.nix:21]: It also uses a zip instead of git.

**Pull request merged**